### PR TITLE
feat(udf): support function that returns multiple columns

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR adds returning multiple columns support for both UDF and UDTF.

For scalar functions, the result type should be defined as `struct`. The python function can return a tuple:
```python
@udf(input_types=['BINARY'], result_type='STRUCT<src_ip VARCHAR, dst_ip VARCHAR, src_port SMALLINT, dst_port SMALLINT>')
def extract_tcp_info(tcp_packet: bytes):
    # ...
    return src_addr, dst_addr, src_port, dst_port
```

```sql
# returns a single column when called in select list
dev=> select extract_tcp_info(E'\\x45000034a8a8400040065b8ac0a8000ec0a80001035d20b6d971b900000000080020200493310000020405b4' :: bytea);
          extract_tcp_info           
-------------------------------------
 (192.168.0.14,192.168.0.1,861,8374)
(1 row)

# returns multiple columns with `select ().*`
dev=> select (extract_tcp_info(E'\\x45000034a8a8400040065b8ac0a8000ec0a80001035d20b6d971b900000000080020200493310000020405b4' :: bytea)).*;
    src_ip    |   dst_ip    | src_port | dst_port 
--------------+-------------+----------+----------
 192.168.0.14 | 192.168.0.1 |      861 |     8374
(1 row)
```

For table functions:
```python
@udtf(input_types=['BINARY'], result_types=['VARCHAR', 'VARCHAR', 'SMALLINT', 'SMALLINT'])
def extract_tcp_info(tcp_packet: bytes):
    # ...
    yield src_addr, dst_addr, src_port, dst_port
```

```sql
# returns a single column when called in select list
dev=> select extract_tcp_info(E'\\x45000034a8a8400040065b8ac0a8000ec0a80001035d20b6d971b900000000080020200493310000020405b4' :: bytea);
          extract_tcp_info           
-------------------------------------
 (192.168.0.14,192.168.0.1,861,8374)
(1 row)

# returns multiple columns when called in from-clause
dev=> select * from extract_tcp_info(E'\\x45000034a8a8400040065b8ac0a8000ec0a80001035d20b6d971b900000000080020200493310000020405b4' :: bytea);
    src_ip    |   dst_ip    | src_port | dst_port 
--------------+-------------+----------+----------
 192.168.0.14 | 192.168.0.1 |      861 |     8374
(1 row)
```

Some discussions:
- Should `extract_tcp_info` be defined as a table function? or a scalar function that returns struct type?
- Can the current implementation be called table function? or are they just [set returning functions](https://www.postgresql.org/docs/current/functions-srf.html)?

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- SQL commands, functions, and operators

### Release note

User-defined table functions can return multiple columns now.

</details>

### CODE REVIEW

**Comment by xxchan** [comment]
> > Can the current implementation be called table function? or are they just [set returning functions](https://www.postgresql.org/docs/current/functions-srf.html)?
> 
> I think it is the `set returning functions`. Because our current table function implementation does not give a way to notice the input is over.

How do you define "table function" here? In my dictionary, TF = SRF 😅

**Comment by st1page** [comment]
> > > Can the current implementation be called table function? or are they just [set returning functions](https://www.postgresql.org/docs/current/functions-srf.html)?
> > 
> > 
> > I think it is the `set returning functions`. Because our current table function implementation does not give a way to notice the input is over.
> 
> How do you define "table function" here? In my dictionary, TF = SRF sweat_smile

just like,
https://www.postgresql.org/docs/7.3/xfunc-tablefunctions.html
https://cloud.google.com/bigquery/docs/reference/standard-sql/table-functions
and the "M to N" table function described by https://docs.snowflake.com/en/sql-reference/functions-table#functions-in-which-each-output-row-depends-upon-multiple-input-rows

in that case, the output rows can depend on any row of the input, so it might can only output the result when all the input comes. e,g, implement a table function that can sort the original input relation with some keys and assign the rank to a new column. it means that
1. it is not friendly for our current UDF design, the udf processing can not be pipelined parallelized and the UDF server should buffer all input data.
2. the function can only be defined in limited input and we must notice the function if the input has been over.  So our streaming execution can not (maybe with a watermark in the future)

**Comment by xxchan** [comment]
> > Didn't think much about it, but "M-N" TF sounds just like OVER windowing?
> 
> I think no, at least the over window function will give the same number of output rows with the input 🤔 but the m-n TF do not have that restriction

Do you have examples for m-n TF?

**Comment by xxchan** [comment]
BTW, snowflake UDTF is only 1-n 🤔

> A tabular function (UDTF) returns a tabular value **for each input row**. In the handler for a UDTF, you write methods that conform to an interface required by Snowflake. 

https://docs.snowflake.com/en/sql-reference/udf-overview#scalar-and-tabular-functions

**Comment by xxchan** [comment]
BTW, our discussion might be off-topic in the PR 🤣

**Comment by st1page** [comment]
> Do you have examples for m-n TF?

In fact, Any standard or not standard relational operator. after some investigation I think all system I know implement it with sql :thinking:

File: e2e_test/udf/test.py [hunk 1]
```diff
@@ -22,16 +24,25 @@ def gcd3(x: int, y: int, z: int) -> int:
     return gcd(gcd(x, y), z)
 
 
+@udf(input_types=['BINARY'], result_type='STRUCT<src_ip VARCHAR, dst_ip VARCHAR, src_port SMALLINT, dst_port SMALLINT>')
```
Comment (TennyZhuang) [inline e2e_test/udf/test.py:27]: 🤔 Is it possible to use Python3 type hints?

Comment (wangrunji0408) [reply e2e_test/udf/test.py:27]: Do you mean to infer SQL type from function signatures?

Comment (TennyZhuang) [reply e2e_test/udf/test.py:27]: yep, just like how `@dataclass` do

Comment (wangrunji0408) [reply e2e_test/udf/test.py:27]: Interesting idea. It seems feasible.

I can think of one more benefit: a python function can be compatible with multiple SQL functions.
For example: `def gcd(x: int, y: int): int` can be used as SQL functions:
- `gcd(INT2, INT2) -> INT2`
- `gcd(INT4, INT4) -> INT4`
- `gcd(INT8, INT8) -> INT8`
- ...

**Review by TennyZhuang — APPROVED** [review]

**Pull request merged**